### ISSUE TITLE
FlaxStableDiffusionXLPipeline outputs 'NAN' in fp16 inference.

### ISSUE THREAD

**Issue by yisonzhu** [huggingface/diffusers#7765]
### Describe the bug

The data of the image generated by FlaxStableDiffusionXLPipeline is all 'nan' with fp16 datatype, as displayed by the following pseudocode:
```
pipeline, params = FlaxStableDiffusionXLPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", dtype=jnp.float16)
images = pipeline(prompt_ids, params).images
print(images)
```
the printed results are all 'nan'. 

If I run with fp32 or bf16 datatype, the accuracy is fine. If I only set the VAE to fp32 and leave the other parts of the model in fp16, the accuracy is also fine.

The reproduction script is from **examples/research_projects/sdxl_flax/sdxl_single.py**, just add a parameter `dtype=jnp.float16` to `FlaxStableDiffusionXLPipeline.from_pretrained()` function and change the params casting type from `jnp.bfloat16` to `jnp.float16`.

Besides, I find `dtype `was not passed into the `FlaxStableDiffusionXLPipeline.from_pretrained()` in **examples/research_projects/sdxl_flax/sdxl_single.py**. I think `dtype `should be specified as its parameter, otherwise all modules contained in the pipeline will receive the "None" `dtype ` and finally all uses the default "fp32" datatype.

### Reproduction

```
# Show best practices for SDXL JAX
import time

import jax
import jax.numpy as jnp
import numpy as np
from flax.jax_utils import replicate

from diffusers import FlaxStableDiffusionXLPipeline

NUM_DEVICES = jax.device_count()

# 1. Let's start by downloading the model and loading it into our pipeline class
# Adhering to JAX's functional approach, the model's parameters are returned seperatetely and
# will have to be passed to the pipeline during inference
pipeline, params = FlaxStableDiffusionXLPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", revision="refs/pr/95", split_head_dim=True, dtype=jnp.float16
)

# 2. We cast all parameters to float16 EXCEPT the scheduler which we leave in
# float32 to keep maximal precision
scheduler_state = params.pop("scheduler")
params = jax.tree_util.tree_map(lambda x: x.astype(jnp.float16), params)
params["scheduler"] = scheduler_state

# 3. Next, we define the different inputs to the pipeline
prompt = "a colorful photo of a castle in the middle of a forest with trees and bushes, by Ismail Inceoglu, shadows, high contrast, dynamic shading, hdr, detailed vegetation, digital painting, digital drawing, detailed painting, a detailed digital painting, gothic art, featured on deviantart"
neg_prompt = "fog, grainy, purple"
default_seed = 33
default_guidance_scale = 5.0
default_num_steps = 25


# 4. In order to be able to compile the pipeline
# all inputs have to be tensors or strings
# Let's tokenize the prompt and negative prompt
def tokenize_prompt(prompt, neg_prompt):
    prompt_ids = pipeline.prepare_inputs(prompt)
    neg_prompt_ids = pipeline.prepare_inputs(neg_prompt)
    return prompt_ids, neg_prompt_ids


# 5. To make full use of JAX's parallelization capabilities
# the parameters and input tensors are duplicated across devices
# To make sure every device generates a different image, we create
# different seeds for each image. The model parameters won't change
# during inference so we do not wrap them into a function
p_params = replicate(params)


def replicate_all(prompt_ids, neg_prompt_ids, seed):
    p_prompt_ids = replicate(prompt_ids)
    p_neg_prompt_ids = replicate(neg_prompt_ids)
    rng = jax.random.PRNGKey(seed)
    rng = jax.random.split(rng, NUM_DEVICES)
    return p_prompt_ids, p_neg_prompt_ids, rng


# 6. Let's now put it all together in a generate function
def generate(
    prompt,
    negative_prompt,
    seed=default_seed,
    guidance_scale=default_guidance_scale,
    num_inference_steps=default_num_steps,
):
    prompt_ids, neg_prompt_ids = tokenize_prompt(prompt, negative_prompt)
    prompt_ids, neg_prompt_ids, rng = replicate_all(prompt_ids, neg_prompt_ids, seed)
    images = pipeline(
        prompt_ids,
        p_params,
        rng,
        num_inference_steps=num_inference_steps,
        neg_prompt_ids=neg_prompt_ids,
        guidance_scale=guidance_scale,
        jit=True,
    ).images

    print(images) # all is nan

    # convert the images to PIL
    images = images.reshape((images.shape[0] * images.shape[1],) + images.shape[-3:])
    return pipeline.numpy_to_pil(np.array(images))


start = time.time()
images = generate(prompt, neg_prompt)

print(f"Inference in {time.time() - start}")

for i, image in enumerate(images):
    image.save(f"castle_{i}.png")

```

### Logs

```shell
[[[[[nan nan nan]
    [nan nan nan]
    [nan nan nan]
    ...
    [nan nan nan]
    [nan nan nan]
    [nan nan nan]]

   [[nan nan nan]
    [nan nan nan]
    [nan nan nan]
    ...
    [nan nan nan]
    [nan nan nan]
    [nan nan nan]]

   [[nan nan nan]
    [nan nan nan]
    [nan nan nan]
    ...
    [nan nan nan]
    [nan nan nan]
    [nan nan nan]]

   ...

   [[nan nan nan]
    [nan nan nan]
    [nan nan nan]
    ...
    [nan nan nan]
    [nan nan nan]
    [nan nan nan]]

   [[nan nan nan]
    [nan nan nan]
    [nan nan nan]
    ...
    [nan nan nan]
    [nan nan nan]
    [nan nan nan]]

   [[nan nan nan]
    [nan nan nan]
    [nan nan nan]
    ...
    [nan nan nan]
    [nan nan nan]
    [nan nan nan]]]]]
```

### System Info

- Platform: Linux
- CUDA Version: 12.3
- python: Python 3.10.12
- jax: 0.4.24
- jaxlib: 0.4.24+cuda12.cudnn89
- diffusers: 0.26.3
- transformers: 4.37.0
- flax: 0.8.1

### Who can help?

@pcuenca

**Comment by standardAI** [comment]
Could you see [here](https://huggingface.co/madebyollin/sdxl-vae-fp16-fix)?

**Comment by yisonzhu** [comment]
> Could you see [here](https://huggingface.co/madebyollin/sdxl-vae-fp16-fix)?

Thanks, it is fixed. It can be also applied to Flax framework.
I will close this issue.
